agent/util-scripts/pbench-tool-meister-start

```diff
@@ -244,13 +204,635 @@ def waitpid(pid):
+    if successes != tm_count:
+        raise StartTmsErr(
+            f"number of created Tool Meisters, {successes}, does not"
+            f" match the expected number of Tool Meisters, {tm_count}",
+            ReturnCode.TMMISSING,
+        )
+
+
+class BaseServer:
+    """BaseServer - abstract base class for common code shared between the
+    ToolDataSink and RedisServer classes.
+    """
+
+    envi_var = None
+    def_port = None
+    bad_port_code = None
+
+    class Err(ReturnCodeErr):
+        pass
+
+    def __init__(self):
+        # Host / port pair for how to connect to the given instance of a
+        # "server".
+        self.host = None
+        self.port = None
+        # (Optional) Host / port pair for how a server will "bind" to offer its
+        # services.
+        self.bind_host = None
+        self.bind_port = None
+        # Whether or not this is instance is managed by the execution of
+        # tool-meister-start, or if it is managed externally.
+        self._managed = None
+
+    def __repr__(self):
+        return f"{self.host}:{self.port} / {self.bind_host}:{self.bind_port} - {self._managed}"
+
+    def __str__(self):
+        return f"{self.host}:{self.port}"
+
+    def managed(self):
+        return self._managed
+
+    def determine_ip_and_port(self, full_hostname):
+        """determine_ip_and_port - from the environment variable specified in
+        envi_var, determine the IP:port for binding (listening) and the IP:port
+        for connecting.
+
+        If the environment variable does not exist, use defaults.
+
+        If the environment variable does exist, it will be in one of two forms:
+
+          - `<ip>:<port>'
+            * where the same ip address and port are used for binding and
+              connecting
+          - `<bind ip>:<port>;<connect ip>:<port>`
+            * where a semi-colon separates the bind ip/port from the connecting
+              ip/port
+
+        In either case, a missing port (bare colon, optional) indicates the
+        default port should be used. If not IP address is given, the full
+        hostname is used.
+
+        No attempt is made to verify that the IP address resolves, or that it
+        is reachable, though we do check they are syntactically valid.
+        """
+        assert (
+            self._managed is True
+            and self.host is None
+            and self.port is None
+            and self.bind_host is None
+            and self.bind_port is None
+        ), f"Logic bomb!  Unexpected state: {self!r}"
+        assert self.envi_var is not None, "Logic bomb!  No envi var set"
+        assert self.def_port is not None, "Logic bomb!  No default port set"
+
+        envi_val = os.environ.get(self.envi_var, full_hostname)
+        parts = envi_val.split(";", 1)
+        pairs = []
+        for part in parts:
+            host_port_parts = part.rsplit(":", 1)
+            if len(host_port_parts) == 1:
+                port = self.def_port
```

Comment:
I like that idea.

>         (host, port) = matches.group(1, 2)

And for what it's worth, I'm a fan of using named capture groups: it makes both the Regex and the code using it a lot more easily readable.

Reply:
This seems much more complicated as a regex than its worth.  We could consider the use of the `ipaddress` module.

Reply:
The regex does the "recognition" (i.e., superficial validation) as well as the parsing in one swell foop, so, in the sum of things, I don't think the regex is really more complicated.  Yeah, it's scary looking, but we can declare the regex at the top of the class where no one will have to look at it.  ;-)

I would certainly support using `ipaddress` as well, but I didn't get the impression that it would handle a FQDN -- I thought it only did numerics.  If it's a full replacement for the regex, then I'm on-board.

Reply:
Added a separate initial commit for general hostname validation, then added an additional commit to use it in the new `pbench-tool-meister-start` code.

Reply:
Moved to PR #2207.